pkg/controller/reconciler.go

```diff
@@ -68,19 +68,36 @@ func (o *operationAnnotationWrapper) Reconcile(request reconcile.Request) (recon
 		return reconcile.Result{}, err
 	}
 
+	if ok, err := o.removeOperationAnnotation(obj, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationReconcile); !ok {
+		return reconcile.Result{}, err
+	}
+
+	result, err := o.Reconciler.Reconcile(request)
+
+	if err == nil {
+		if ok, err := o.removeOperationAnnotation(obj, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationMigrate); !ok {
```

Comment:
I think removing the `migrate` annotation after the reconciliation doesn't work with the current implementation. Let's say you start the reconciliation (`o.Reconciler.Reconcile(request)`) which also updates the resource a bunch of times. While sending the updates, the resource still carries the `reconcile` operation. As a consequence, the triggered update enqueues the resource again because `HasOperationAnnotation()` returns true, i.e. the resource is infinitely re-queued.

Have you tested this scenario?

Reply:
Generally for the OperationAnnotationWrapper, which is used by all controllers, the `restore` annotation should be handled exactly the same as the `reconcile` annotation

Reply:
> Meant to say "all controllers which do not have to restore their state". Basically during migration Gardener doesn't know whether an extension controller has to restore its state or simply perform reconciliation on the new seed, so it sets operation=restore on all extension resources.
Extension controllers which do not have to restore their state, simply have to remove the annotation and perform reconciliation. Whereas controllers which have to restore their state need to trigger restoration instead of reconciliation.

So what will be done is that a new struct similar to OperationAnnotationWrapper will be created which should handle the restore and reconcile operations differently.

Sorry I commented the wrong conversation.

Okay, makes sense what you say with an additional wrapping functionality.

Reply:
We shall move the Restore handling in another Wrapper.